# Exemplo de regressão linear múltipla: exercício soja

Primeiro precisamos chamar os dados para nosso projeto. Meus dados estarão 'embeded' neste code.

# Dados

In [2]:
import pandas as pd
dados = pd.DataFrame.from_dict(
  {'OBS': {0: 1, 1: 2, 2: 3, 3: 4, 4: 5, 5: 6, 6: 7, 7: 8, 8: 9, 9: 10, 10: 11, 11: 12, 12: 13, 13: 14, 14: 15, 15: 16, 
           16: 17, 17: 18, 18: 19, 19: 20, 20: 21, 21: 22, 22: 23, 23: 24, 24: 25, 25: 26, 26: 27, 27: 28, 28: 29, 29: 30, 
           30: 31, 31: 32, 32: 33, 33: 34, 34: 35, 35: 36, 36: 37, 37: 38, 38: 39, 39: 40, 40: 41, 41: 42, 42: 43, 43: 44, 
           44: 45, 45: 46, 46: 47, 47: 48, 48: 49, 49: 50, 50: 51, 51: 52, 52: 53, 53: 54, 54: 55, 55: 56, 56: 57, 57: 58, 
           58: 59, 59: 60, 60: 61, 61: 62, 62: 63, 63: 64, 64: 65, 65: 66, 66: 67, 67: 68, 68: 69, 69: 70, 70: 71, 71: 72, 
           72: 73, 73: 74, 74: 75, 75: 76, 76: 77, 77: 78, 78: 79, 79: 80, 80: 81, 81: 82, 82: 83, 83: 84, 84: 85, 85: 86, 
           86: 87, 87: 88, 88: 89, 89: 90, 90: 91, 91: 92, 92: 93, 93: 94, 94: 95, 95: 96, 96: 97, 97: 98, 98: 99, 99: 100, 
           100: 101, 101: 102, 102: 103, 103: 104, 104: 105, 105: 106, 106: 107, 107: 108, 108: 109, 109: 110, 110: 111, 
           111: 112, 112: 113, 113: 114, 114: 115, 115: 116, 116: 117}, 
   'QSOJA': {0: 436.631327347, 1: 373.648319403, 2: 394.422208122, 3: 343.569529223, 4: 303.766149519, 5: 301.164159253, 
             6: 288.948162961, 7: 330.653425923, 8: 312.790481897, 9: 326.337437514, 10: 393.924244131, 11: 472.095484821, 
             12: 506.519816219, 13: 351.622349614, 14: 381.683735178, 15: 383.16244294, 16: 411.039886175, 17: 393.721292241, 
             18: 434.570723074, 19: 433.61603289, 20: 397.521061235, 21: 392.667303139, 22: 388.161060061, 23: 370.962499467, 
             24: 392.989989558, 25: 364.608287145, 26: 346.432408617, 27: 418.249947335, 28: 406.403616915, 29: 335.565654517, 
             30: 372.389277147, 31: 355.138034335, 32: 350.368666514, 33: 333.22912698, 34: 331.404160354, 35: 350.215587437, 
             36: 347.930917294, 37: 429.353837601, 38: 312.648633868, 39: 320.1290397, 40: 367.600375264, 41: 370.58115319, 
             42: 318.293875369, 43: 360.716491231, 44: 344.127888634, 45: 348.460445231, 46: 339.909909323, 47: 355.115806958, 
             48: 333.991242698, 49: 324.352196839, 50: 326.362748629, 51: 337.522873509, 52: 326.439134587, 53: 315.883680773, 
             54: 309.389262881, 55: 309.992167966, 56: 294.858595183, 57: 319.126938705, 58: 321.075126328, 59: 324.617110436, 
             60: 326.498169984, 61: 323.024096765, 62: 306.607962724, 63: 316.685380598, 64: 306.63234033, 65: 347.051171678, 
             66: 281.018277888, 67: 306.438241825, 68: 310.158071775, 69: 308.554712739, 70: 317.988817729, 71: 309.3024648, 
             72: 301.907326808, 73: 293.695986672, 74: 286.246007121, 75: 284.741951642, 76: 281.541824884, 77: 276.076484065, 
             78: 225.250102468, 79: 221.579142339, 80: 222.819046328, 81: 210.465091286, 82: 204.579726173, 83: 210.208100729, 
             84: 214.619137203, 85: 249.68373735, 86: 234.056997721, 87: 237.782743552, 88: 247.783594823, 89: 243.326935015, 
             90: 250.517759798, 91: 245.477283956, 92: 242.547637962, 93: 235.139515392, 94: 246.077631412, 95: 300.660379261, 
             96: 311.547314244, 97: 311.592498254, 98: 311.661546245, 99: 313.521069724, 100: 324.623216411, 101: 325.219601572, 
             102: 316.051963666, 103: 315.23510561, 104: 313.039404973, 105: 311.256344161, 106: 314.759829619, 107: 319.859862035, 
             108: 315.86486682, 109: 313.067146865, 110: 305.235250016, 111: 299.911393983, 112: 292.819273066, 113: 288.374750217, 
             114: 282.574142328, 115: 280.040196223, 116: 272.093598783}, 
   'FERTILIZANTE': {0: 19.0271541214, 1: 17.896131535, 2: 16.7816326404, 3: 13.4907436954, 4: 9.8792199643, 5: 9.47578570764, 
                    6: 11.3642792008, 7: 15.1279345194, 8: 15.3328667597, 9: 12.851502126, 10: 11.5137639555, 11: 12.855099231, 
                    12: 13.0130463524, 13: 13.4551480743, 14: 14.3478259384, 15: 13.4461834272, 16: 12.8505836532, 17: 11.878680085, 
                    18: 8.97428388969, 19: 11.2853667097, 20: 10.3459526645, 21: 10.1678109845, 22: 12.599515057, 23: 18.0635955859, 
                    24: 22.5905514861, 25: 27.0789975599, 26: 25.8933999151, 27: 24.670302157, 28: 22.3482879799, 29: 23.0661021802, 
                    30: 23.46202603, 31: 23.3742807993, 32: 23.6293106114, 33: 21.5085651347, 34: 20.4244350802, 35: 18.8640383209, 
                    36: 18.2649424041, 37: 16.1258445984, 38: 15.3872613296, 39: 15.7475620647, 40: 14.8983557849, 41: 18.5713738614, 
                    42: 19.782502768, 43: 19.9334418708, 44: 20.8217736318, 45: 20.2664704198, 46: 18.7039332243, 47: 16.9397047639, 
                    48: 15.2224405941, 49: 15.3900606317, 50: 15.4866979181, 51: 14.1888364722, 52: 17.5057959027, 53: 18.4499760981, 
                    54: 18.0443380567, 55: 18.7653974258, 56: 17.733992169, 57: 14.4708212634, 58: 19.1906127545, 59: 23.6107568782, 
                    60: 22.4809389477, 61: 19.9623501966, 62: 24.292193006, 63: 23.7774100871, 64: 20.902854945, 65: 17.6135324226, 
                    66: 18.2422805406, 67: 18.4963907183, 68: 20.1887054098, 69: 17.2164566133, 70: 18.2228548842, 71: 17.63412, 
                    72: 17.3651340732, 73: 21.165752859, 74: 20.6546570465, 75: 20.5382250175, 76: 20.2625821575, 77: 20.0316705895, 
                    78: 19.6747891284, 79: 19.2320746157, 80: 19.1556951795, 81: 18.6659751037, 82: 18.2573530021, 83: 18.0250472242, 
                    84: 18.2220357618, 85: 18.1808546614, 86: 17.9427350427, 87: 17.8936494922, 88: 17.5782913467, 89: 17.4452921577, 
                    90: 17.4075434149, 91: 17.287132673, 92: 17.0010026609, 93: 16.7956796708, 94: 16.6141444443, 95: 16.6135183413, 
                    96: 16.5922588804, 97: 22.0744924554, 98: 22.0123280182, 99: 21.8207871467, 100: 21.4818658909, 101: 21.3918043526, 
                    102: 21.1455500396, 103: 21.0219803014, 104: 20.9587175263, 105: 20.8135574015, 106: 22.8749876176, 107: 22.8850270408, 
                    108: 22.7508743781, 109: 22.6733292656, 110: 22.4866071739, 111: 22.3322073706, 112: 22.1375395445, 113: 22.1331443789, 
                    114: 22.081593902, 115: 22.1113459316, 116: 22.0611011027}, 
   'TRATOR': {0: 3.1712177231, 1: 2.9130726375, 2: 2.79693877341, 3: 2.89345709284, 4: 3.09884379577, 5: 3.55341964037, 6: 3.9774977203, 
              7: 4.86255038122, 8: 5.27067294863, 9: 5.29179499304, 10: 4.38619579258, 11: 3.61549665871, 12: 3.18096688613, 13: 3.31203644906, 
              14: 3.58695648459, 15: 3.36154585679, 16: 2.79360514201, 17: 3.25443289999, 18: 2.69228516691, 19: 2.90195143963, 
              20: 2.95598647557, 21: 2.98255788878, 22: 3.23987530037, 23: 3.18769333868, 24: 3.80472446082, 25: 4.16599962459, 
              26: 4.25644930111, 27: 3.65485957882, 28: 3.42843054238, 29: 3.56476124603, 30: 4.10585455524, 31: 4.23376550096, 
              32: 4.21951975203, 33: 4.45534563505, 34: 4.22104991658, 35: 3.77280766418, 36: 4.05887608981, 37: 4.35397804157, 
              38: 3.71858815465, 39: 3.9194002239, 40: 3.86253116707, 41: 3.4489694314, 42: 3.40698658782, 43: 3.10590838452, 
              44: 3.40719932157, 45: 3.89739815766, 46: 3.58705568686, 47: 3.06139242721, 48: 2.56377946849, 49: 3.27038788423, 
              50: 3.61356284755, 51: 2.9212310384, 52: 3.72808616447, 53: 4.09999468846, 54: 4.00985290148, 55: 3.12756623763, 
              56: 3.66502504826, 57: 3.61770531584, 58: 2.74151610779, 59: 3.07966394064, 60: 2.24809389477, 61: 2.21803891073, 
              62: 3.23895906746, 63: 3.26939388698, 64: 2.71737114285, 65: 3.52270648453, 66: 2.63499607808, 67: 2.84559857204, 
              68: 3.02830581147, 69: 2.75463305812, 70: 3.79642810087, 71: 3.673775, 72: 4.3412835183, 73: 4.23315057181, 
              74: 5.50790854573, 75: 5.47686000466, 76: 5.40335524199, 77: 6.67722352984, 78: 6.5582630428, 79: 6.41069153856, 
              80: 6.3852317265, 81: 6.22199170124, 82: 6.08578433403, 83: 6.00834907473, 84: 6.07401192061, 85: 6.06028488713, 
              86: 5.98091168091, 87: 5.96454983075, 88: 5.85943044889, 89: 5.8150973859, 90: 5.80251447164, 91: 5.76237755766, 
              92: 5.66700088697, 93: 5.59855989028, 94: 5.53804814809, 95: 5.53783944709, 96: 5.53075296013, 97: 5.51862311384, 
              98: 5.50308200454, 99: 5.45519678667, 100: 5.37046647273, 101: 5.34795108814, 102: 5.28638750989, 103: 5.25549507535, 
              104: 5.23967938158, 105: 5.20338935039, 106: 5.19886082219, 107: 5.20114250927, 108: 5.17065326775, 109: 5.15302937855, 
              110: 5.11059253952, 111: 5.07550167513, 112: 6.03751078486, 113: 6.03631210335, 114: 6.02225288236, 115: 6.03036707224, 
              116: 6.01666393711}, 
   'MO': {0: 0.0680761131536, 1: 0.0680761131536, 2: 0.0680761131536, 3: 0.0680761131536, 4: 0.0715237353179, 5: 0.0833559863149, 
          6: 0.0985723372523, 7: 0.111410334113, 8: 0.114487135409, 9: 0.102686021888, 10: 0.121305727388, 11: 0.123194700963, 
          12: 0.113382190903, 13: 0.0988435877764, 14: 0.0807663035114, 15: 0.0622686351567, 16: 0.083800394246, 17: 0.0765243736067, 
          18: 0.0574877668834, 19: 0.0653655605136, 20: 0.0725099376921, 21: 0.0780548956575, 22: 0.0806568955951, 23: 0.0816669323964, 
          24: 0.0822771664653, 25: 0.0801052294483, 26: 0.0750435658726, 27: 0.0703103611476, 28: 0.0796030780377, 29: 0.0830484524405, 
          30: 0.093584156327, 31: 0.100913962297, 32: 0.108669511694, 33: 0.111606408158, 34: 0.118828985067, 35: 0.120133794043, 
          36: 0.118234214897, 37: 0.105740298613, 38: 0.093457488034, 39: 0.0841934657141, 40: 0.0798233440708, 41: 0.0747407520408, 
          42: 0.0708495225001, 43: 0.0672990598108, 44: 0.0611103807825, 45: 0.055394759321, 46: 0.0660366632049, 47: 0.0689438046948, 
          48: 0.0669589093824, 49: 0.0638618759693, 50: 0.0596771157157, 51: 0.0549263770463, 52: 0.073413760912, 53: 0.0823331322263, 
          54: 0.0834334103063, 55: 0.079906745817, 56: 0.0716246518416, 57: 0.0633701381158, 58: 0.0934971222594, 59: 0.104075531949, 
          60: 0.100833255886, 61: 0.0930097649901, 62: 0.0813417334974, 63: 0.0695736926403, 64: 0.21049581374, 65: 0.261260874072, 
          66: 0.258725366517, 67: 0.236599269382, 68: 0.206871841743, 69: 0.167746823366, 70: 0.126327870706, 71: 0.114329213918, 
          72: 0.104788717578, 73: 0.0945762804116, 74: 0.0848737411962, 75: 0.0770183438155, 76: 0.0741085180759, 77: 0.071409196083, 
          78: 0.0683152400292, 79: 0.0649972892104, 80: 0.0629654795252, 81: 0.0596274204703, 82: 0.0578149511733, 83: 0.0565786204537, 
          84: 0.0566907779257, 85: 0.0560576352059, 86: 0.0548250237417, 87: 0.0541779942959, 88: 0.0539450091897, 89: 0.0542532433056, 
          90: 0.0548506857575, 91: 0.0551811677564, 92: 0.0549659731863, 93: 0.0549918545222, 94: 0.0547197615699, 95: 0.0550399709491, 
          96: 0.0552913982385, 97: 0.0554912883605, 98: 0.0556552669895, 99: 0.0554884433151, 100: 0.0541104333052, 101: 0.0533695807757, 
          102: 0.0522471298894, 103: 0.0514366981903, 104: 0.0507783150735, 105: 0.049926520817, 106: 0.049902565317, 107: 0.0499439709453, 
          108: 0.0496705879533, 109: 0.0495206123279, 110: 0.0491319590269, 111: 0.0488136373605, 112: 0.0483839405954, 113: 0.0483701426115, 
          114: 0.0482533012199, 115: 0.0483141284115, 116: 0.0482001633184}})
dados

OBS       QSOJA  FERTILIZANTE    TRATOR        MO
0      1  436.631327     19.027154  3.171218  0.068076
1      2  373.648319     17.896132  2.913073  0.068076
2      3  394.422208     16.781633  2.796939  0.068076
3      4  343.569529     13.490744  2.893457  0.068076
4      5  303.766150      9.879220  3.098844  0.071524
..   ...         ...           ...       ...       ...
112  113  292.819273     22.137540  6.037511  0.048384
113  114  288.374750     22.133144  6.036312  0.048370
114  115  282.574142     22.081594  6.022253  0.048253
115  116  280.040196     22.111346  6.030367  0.048314
116  117  272.093599     22.061101  6.016664  0.048200

[117 rows x 5 columns]

Estatísticas básicas (descritivas) do dataset:

In [3]:
dados.describe()

OBS       QSOJA  FERTILIZANTE      TRATOR          MO
count  117.000000  117.000000    117.000000  117.000000  117.000000
mean    59.000000  322.254407     18.404439    4.317215    0.081653
std     33.919021   56.012721      3.818477    1.199974    0.039856
min      1.000000  204.579726      8.974284    2.218039    0.048200
25%     30.000000  294.858595     16.592259    3.269394    0.055395
50%     59.000000  316.685381     18.449976    4.058876    0.070850
75%     88.000000  350.215587     21.391804    5.455197    0.093457
max    117.000000  506.519816     27.078998    6.677224    0.261261

Podemos ver a estrutura dos dados.

In [4]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 117 entries, 0 to 116
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   OBS           117 non-null    int64  
 1   QSOJA         117 non-null    float64
 2   FERTILIZANTE  117 non-null    float64
 3   TRATOR        117 non-null    float64
 4   MO            117 non-null    float64
dtypes: float64(4), int64(1)
memory usage: 5.5 KB


# Regressão

Usaremos os pacotes numpy, pandas, sklearn.linear_model, statsmodels.formula.api.

In [5]:
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
import statsmodels.formula.api as sm
eq1 = sm.ols(formula="QSOJA~FERTILIZANTE+TRATOR+MO", data=dados).fit()
print(eq1.summary())

                            OLS Regression Results                            
Dep. Variable:                  QSOJA   R-squared:                       0.465
Model:                            OLS   Adj. R-squared:                  0.451
Method:                 Least Squares   F-statistic:                     32.75
Date:                Sun, 02 Apr 2023   Prob (F-statistic):           2.61e-15
Time:                        20:34:10   Log-Likelihood:                -599.90
No. Observations:                 117   AIC:                             1208.
Df Residuals:                     113   BIC:                             1219.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept      494.9657     25.572     19.356   

Portanto, nosso modelo tem um $R^2 = 0.465$, razoavelmente baixo, indicando que apenas 46% das variações de QSOJA são explicadas pelas variáveis explicativas.  

Os coeficientes do intercepto, de TRATOR e de MO foram estatisticamente significativos, respectivamente a 1%, 1% e 10% de significância. Ou seja, não foi possível dizer que FERTILIZANTE afeta QSOJA.

Podemos agora testar para as pressuposições do modelo de regressão linear clássico.